### PR TITLE
fea(api): manage downtimes for host and service

### PR DESCRIPTION
**Pull request by callapa**
## Description

Allow to add/list/cancel an downtime on host or service

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [x] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x
- [x] 20.04.x (master)

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: src/Centreon/Application/Controller/DowntimeController.php [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+         */
+        $downtime = $serializer->deserialize(
+            (string)$request->getContent(),
+            Downtime::class,
+            'json'
+        );
+
+        $this->downtimeService->addHostDowntime($downtime, $host);
+
+        if ($withService) {
```
Comment (kduret) [inline src/Centreon/Application/Controller/DowntimeController.php]:
```suggestion
        if ($withService === true) {
```

File: src/Centreon/Application/Controller/DowntimeController.php [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+             * @var Downtime $downtime
+             */
+            $downtime = $serializer->deserialize(
+                (string) $request->getContent(),
+                Downtime::class,
+                'json'
+            );
+            $this->monitoringService->filterByContact($contact);
+
+            $host = $this->monitoringService->findOneHost($hostId);
```
Comment (kduret) [inline src/Centreon/Application/Controller/DowntimeController.php]: Do you need to make this check ?
Cause you already check if service (and/or host) exists in the next lines

File: src/Centreon/Application/Controller/DowntimeController.php [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    {
+        $contact = $this->getUser();
+        if ($contact === null) {
+            return $this->view(null, Response::HTTP_UNAUTHORIZED);
+        }
+        $hostsDowntime = $this->downtimeService
+            ->filterByContact($contact)
+            ->findDowntimes();
+
+        $context = (new Context())->setGroups(['dwt_main', 'dwt_service']);
```
Comment (kduret) [inline src/Centreon/Application/Controller/DowntimeController.php]: should not be better to use complete name ? downtime_main and downtime_service

File: src/Centreon/Application/Controller/DowntimeController.php [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+        $host = $this->monitoringService
+            ->filterByContact($contact)
+            ->findOneHost($downtime->getHostId());
+
+        if ($host === null) {
+            return View::create(null, Response::HTTP_NOT_FOUND, []);
+        }
+
+        if (!$contact->isAdmin()) {
+
```
Comment (kduret) [inline src/Centreon/Application/Controller/DowntimeController.php]:
```suggestion
```

File: src/Centreon/Infrastructure/Downtime/DowntimeRepositoryRDB.php [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+            FROM `:dbstg`.downtimes dwt
+            INNER JOIN `:dbstg`.hosts
+                ON hosts.host_id = dwt.host_id
+            INNER JOIN `:dbstg`.services srv
+                ON srv.service_id = dwt.service_id
+                AND srv.host_id = hosts.host_id
+                AND dwt.host_id = :host_id
+                AND srv.service_id = :service_id
+            INNER JOIN `:db`.`contact`
+                ON contact.contact_alias = dwt.author'
```
Comment (kduret) [inline src/Centreon/Infrastructure/Downtime/DowntimeRepositoryRDB.php]: If contact is deleted from centreon, we should be able to get the downtime information

File: src/Centreon/Application/Controller/DowntimeController.php (hunk 2 above)
Comment (callapa) [reply src/Centreon/Application/Controller/DowntimeController.php]: You're right

File: src/Centreon/Application/Controller/DowntimeController.php (hunk 3 above)
Comment (callapa) [reply src/Centreon/Application/Controller/DowntimeController.php]: Why not

**Pull request merged**